netket/graph/symmetry.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+
+import numpy as np
+
+from . import AbstractGraph
+from netket.utils.semigroup import SemiGroup
+
+
+@dataclass
+class SymmGroup(SemiGroup):
+    """
+    Collection of symmetry operations acting on the sites of a graph
+    (graph automorphisms).
+    """
+
+    graph: AbstractGraph
+    """Underlying graph"""
+
+    def __matmul__(self, other):
+        if isinstance(other, SymmGroup) and self.graph is not other.graph:
```

Comment:
```suggestion
        if isinstance(other, SymmGroup) and self.graph != other.graph:
```

Reply:
(if that does not work I'll make a PR to fix it in graph side rather than here.)

Reply:
Well, graphs do have `__eq__`, but maybe not in the desired fashion:
```python
>>> 
nk.graph.Chain(8) != nk.graph.Chain(8)
True
```

Reply:
Yes, because we have not implemented it yet and the default `==` of python classes falls back to `is` (I think). 
What I am proposing right now should work identically to before, but if ever we implement a proper __eq__ for graph, it will behave better.

Reply:
Yes, agreed.